shanoir-ng-front/src/app/datasets/shared/dataset.service.ts

```diff
@@ -81,10 +81,13 @@ export class DatasetService extends EntityService<Dataset> implements OnDestroy
             case HttpEventType.Response:
                 progressBar.progress = 0;
                 saveAs(event.body, this.getFilename(event));
+                if (callback) {
+                    callback();
+                }
         }
     }
 
-    public downloadDatasets(ids: number[], format: string, progressBar: LoadingBarComponent) {
+    public downloadDatasets(ids: number[], format: string, progressBar: LoadingBarComponent, callback?) {
```

Comment:
I think you should use a Promise here. Actually promise are essentially callbacks but with a "cleaner" syntax. There might be a difficulty here considering we deal with (righfuly) an observable. But after reading the doc, toPromise() triggers on the completion of the observable. (complete event can be used as the 3rd argument in subscribe(), after error => {}). So I would do that : 

```
public downloadDatasets(ids: number[], format: string, progressBar: LoadingBarComponent): Promise<HttpEvent<any>> {
        const formData: FormData = new FormData();
        formData.set('datasetIds', ids.join(","));
        formData.set("format", format);
        
        let postResponse: Observable<HttpEvent<any>> = this.http.post(
            AppUtils.BACKEND_API_DATASET_URL + '/massiveDownload', formData, {
                reportProgress: true,
                observe: 'events',
                responseType: 'blob'
        });
        this.subscribtions.push(
            postResponse.subscribe(
                (event: HttpEvent<any>) => this.progressBarFunc(event, progressBar),
                error =>  {
                    this.errorService. handleError(error);
                    progressBar.progress = 0;
                }
            )
        );
        return postResponse.toPromise();
    }
```